lib-ee/emqx_license/test/emqx_license_checker_SUITE.erl

```diff
@@ -0,0 +1,179 @@
+%% Tests
+%%------------------------------------------------------------------------------
+
+t_dump(_Config) ->
+    License = mk_license(
+                ["220111",
+                 "0",
+                 "10",
+                 "Foo",
+                 "contact@foo.com",
+                 "20220111",
+                 "100000",
+                 "10"]),
+
+    #{} = emqx_license_checker:update(License),
+
+    ?assertEqual(
+       [{customer,<<"Foo">>},
+        {email,<<"contact@foo.com">>},
+        {max_connections,10},
+        {start_at,<<"2022-01-11">>},
+        {expiry_at,<<"2295-10-27">>},
+        {type,<<"trial">>},
+        {customer_type,10},
+        {expiry,false}],
+       emqx_license_checker:dump()).
+
+t_update(_Config) ->
+    License = mk_license(
+                ["220111",
+                 "0",
+                 "10",
+                 "Foo",
+                 "contact@foo.com",
+                 "20220111",
+                 "100000",
+                 "123"]),
+    #{} = emqx_license_checker:update(License),
+
+    ?assertMatch(
+       #{max_connections := 123},
+       emqx_license_checker:limits()).
+
+t_expired_trial(_Config) ->
+    {NowDate, _} = calendar:universal_time(),
+    Date10DaysAgo = calendar:gregorian_days_to_date(
+                      calendar:date_to_gregorian_days(NowDate) - 10),
+
+    License = mk_license(
+                ["220111",
+                 "0",
+                 "10",
+                 "Foo",
+                 "contact@foo.com",
+                 format_date(Date10DaysAgo),
+                 "1",
+                 "123"]),
+    #{} = emqx_license_checker:update(License),
+
+    ?assertMatch(
+       #{max_connections := 0},
+       emqx_license_checker:limits()).
+
+t_overexpired_small_client(_Config) ->
+    {NowDate, _} = calendar:universal_time(),
+    Date100DaysAgo = calendar:gregorian_days_to_date(
+                      calendar:date_to_gregorian_days(NowDate) - 100),
+
+    License = mk_license(
+                ["220111",
+                 "1",
+                 "0",
+                 "Foo",
+                 "contact@foo.com",
+                 format_date(Date100DaysAgo),
+                 "1",
+                 "123"]),
+    #{} = emqx_license_checker:update(License),
+
+    ?assertMatch(
+       #{max_connections := 0},
+       emqx_license_checker:limits()).
+
+t_overexpired_medium_client(_Config) ->
+    {NowDate, _} = calendar:universal_time(),
+    Date100DaysAgo = calendar:gregorian_days_to_date(
+                      calendar:date_to_gregorian_days(NowDate) - 100),
+
+    License = mk_license(
+                ["220111",
+                 "1",
+                 "1",
+                 "Foo",
+                 "contact@foo.com",
+                 format_date(Date100DaysAgo),
+                 "1",
+                 "123"]),
+    #{} = emqx_license_checker:update(License),
+
+    ?assertMatch(
+       #{max_connections := 123},
```

Comment:
Q:  the licenses of medium+ clients never expire?  :speak_no_evil:

Reply:
Yes, this is the inherited logic. Maybe such clients are controlled legally, and this logic allows them not to suffer from some kinds of internal bureaucratic delays.